src/account_manager.rs

```diff
@@ -998,44 +999,69 @@ async fn poll(accounts: AccountStore, storage_file_path: PathBuf, is_mqtt_monito
                     .collect();
-                (account_id, unconfirmed_messages)
+                (account_handle.clone(), unconfirmed_messages)
             };
 
-            let mut promotions = vec![];
-            let mut reattachments = vec![];
+            let mut reattachments = Vec::new();
+            let mut promotions = Vec::new();
+            let mut no_need_promote_or_reattach = Vec::new();
             for (message_id, payload) in unconfirmed_messages {
-                let new_message = repost_message(account_handle.clone(), &message_id, RepostAction::Retry).await?;
-                if new_message.payload() == &payload {
-                    reattachments.push(new_message);
-                } else {
-                    log::info!("[POLLING] promoted and new message is {:?}", new_message.id());
-                    promotions.push(new_message);
+                match repost_message(account_handle.clone(), &message_id, RepostAction::Retry).await {
+                    Ok(new_message) => {
+                        if new_message.payload() == &payload {
+                            reattachments.push(new_message);
+                        } else {
+                            log::info!("[POLLING] promoted and new message is {:?}", new_message.id());
+                            promotions.push(new_message);
+                        }
+                    }
+                    Err(crate::Error::ClientError(iota::client::Error::NoNeedPromoteOrReattach(_))) => {
+                        no_need_promote_or_reattach.push(message_id);
+                    }
+                    _ => {}
                 }
             }
 
             retried_messages.push(RetriedData {
                 promoted: promotions,
                 reattached: reattachments,
-                account_id,
+                no_need_promote_or_reattach,
+                account_handle,
             });
         }
 
         retried_messages
     };
 
-    retried.iter().for_each(|retried_data| {
-        retried_data.reattached.iter().for_each(|message| {
-            emit_transaction_event(TransactionEventType::Reattachment, &retried_data.account_id, &message);
-        });
-    });
+    for retried_data in retried {
+        let mut account = retried_data.account_handle.write().await;
+        let client = crate::client::get_client(account.client_options()).await;
+        let account_id = account.id();
+
+        for message in &retried_data.reattached {
+            emit_transaction_event(TransactionEventType::Reattachment, &account_id, &message);
+        }
+
+        account.append_messages(retried_data.reattached);
+        account.append_messages(retried_data.promoted);
+
+        for message_id in retried_data.no_need_promote_or_reattach {
+            let message = account.get_message_mut(&message_id).unwrap();
+            if let Ok(metadata) = client.read().await.get_message().metadata(&message_id).await {
```

Comment:
You shouldn't assume that it's confirmed here, if you send a message and call get metadata it could return that reattachment and promotion isn't needed and then still get left behind

Reply:
in this case, isn't the message unconfirmed?

Reply:
Yes